### PR TITLE
prevent offer starvation

### PR DESCRIPTION
**Pull request by felixb**
this change prevents offer starvation by:

* rejecting offers for longer period if no slave is queued
* calling reviveOffers when adding a slave to queue

this is necessary when running a lot of frameworks on a mesos cluster.
the rejectOfferDuration is configurable and defaults to 10min.

this is basically the same change made in marathon and chronos a few weeks ago.
see https://github.com/mesosphere/marathon/pull/1949 for further details.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 1]
```diff
@@ -170,6 +170,11 @@ public synchronized boolean isRunning() {
   public synchronized void requestJenkinsSlave(Mesos.SlaveRequest request, Mesos.SlaveResult result) {
     LOGGER.info("Enqueuing jenkins slave request");
     requests.add(new Request(request, result));
+    if (driver != null) {
+      // ask mesos to send all offers, even the those we declined earlier
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: We capitalize our sentences and end them with periods. Can you please fix here and elsewhere.

s/ask/Ask/
s/details/details./

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 2]
```diff
@@ -261,6 +266,19 @@ public void disconnected(SchedulerDriver driver) {
   public synchronized void resourceOffers(SchedulerDriver driver, List<Offer> offers) {
     LOGGER.fine("Received offers " + offers.size());
     for (Offer offer : offers) {
+      if (requests.isEmpty()) {
+        // decline offer for a longer period if no slave is waiting to get spawned
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Capitalization.

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 3]
```diff
@@ -261,6 +266,19 @@ public void disconnected(SchedulerDriver driver) {
...
+        // this prevents getting offers every few seconds
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: how about:

This prevents unnecessarily getting offers every few seconds and causing starvation when running a lot of frameworks.

You can kill the lines about marathon and chronos.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 4]
```diff
@@ -53,6 +53,8 @@
 import org.kohsuke.stapler.StaplerRequest;
 
 public class MesosCloud extends Cloud {
+  private static final String DEFAULT_REJECT_OFFER_DURATION = "600000";
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: is this in ms? to clarify you can do 

private static final String DEFAULT_REJECT_OFFER_DURATION_MS = "600000"; // 10 mins.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 5]
```diff
@@ -63,6 +65,7 @@
   private final boolean checkpoint; // Set true to enable checkpointing. False by default.
   private boolean onDemandRegistration; // If set true, this framework disconnects when there are no builds in the queue and re-registers when there are.
   private String jenkinsURL;
+  private String rejectOfferDuration;
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: s/reject/decline/

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 6]
```diff
@@ -398,6 +403,40 @@ public void setJenkinsURL(String jenkinsURL) {
...
+  }
+
+  public double getRejectOfferDurationDouble() {
+    return Double.parseDouble(getRejectOfferDuration());
+  }
+
+  public void setRejectOfferDuration(String rejectOfferDuration) {
+    try {
+      if (rejectOfferDuration == null) {
+        LOGGER.info("Missing rejectOfferDuration. Using default " + DEFAULT_REJECT_OFFER_DURATION + ".");
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: please log the units of duration.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 7]
```diff
@@ -123,7 +126,8 @@ public MesosCloud(
       List<MesosSlaveInfo> slaveInfos,
       boolean checkpoint,
       boolean onDemandRegistration,
-      String jenkinsURL) throws NumberFormatException {
+      String jenkinsURL,
+      String rejectOfferDuration) throws NumberFormatException {
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: s/reject/decline/

here and everywhere else in this file.

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly [hunk 8]
```diff
@@ -45,6 +45,11 @@
                     <f:radio name="onDemandRegistration" value="false" checked="${instance.onDemandRegistration == false}" id="onDemandRegistration.false"/>
                 <st:nbsp/>${%No}
         </f:entry>
+
+        <f:entry title="${%Reject offer duration}" field="rejectOfferDuration" description="${%Decline offers for specified amount of time in case no slave is queued. The plugin asks for new offers in case a new slave needs to be started. In milliseconds.}">
```
Comment (vinodkone) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly]: s/Reject/Decline/

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 9]
```diff
@@ -398,6 +403,40 @@ public void setJenkinsURL(String jenkinsURL) {
...
+  }
+
+  public void setRejectOfferDuration(String rejectOfferDuration) {
+    try {
+      if (rejectOfferDuration == null) {
+        LOGGER.info("Missing rejectOfferDuration. Using default " + DEFAULT_REJECT_OFFER_DURATION + ".");
+        this.rejectOfferDuration = DEFAULT_REJECT_OFFER_DURATION;
+      } else {
+        double duration = Double.parseDouble(rejectOfferDuration);
+        if (duration >= 1000) {
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: is it possible for the user to set a value < 1000 given the "min" condition in config.jelly?

File: src/test/java/org/jenkinsci/plugins/mesos/JenkinsSchedulerTest.java [hunk 10]
```diff
@@ -213,47 +255,52 @@ public void testConstructMesosCommandInfoWithNullCustomDockerShell() throws Exce
...
-                    "test-docker-in-docker-image",
-                    Boolean.TRUE,
-                    Boolean.TRUE,
-                    useCustomDockerCommandShell,
-                    customDockerCommandShell,
-                    Collections.<MesosSlaveInfo.Volume>emptyList(),
-                    Collections.<MesosSlaveInfo.Parameter>emptyList(),
-                    Protos.ContainerInfo.DockerInfo.Network.HOST.name(),
-                    Collections.<MesosSlaveInfo.PortMapping>emptyList());
+            "docker",
```
Comment (vinodkone) [inline src/test/java/org/jenkinsci/plugins/mesos/JenkinsSchedulerTest.java]: why the change in indentation here?

File: src/test/java/org/jenkinsci/plugins/mesos/JenkinsSchedulerTest.java [hunk 11]
```diff
@@ -213,47 +255,52 @@ public void testConstructMesosCommandInfoWithNullCustomDockerShell() throws Exce
...
+        return new Mesos.SlaveRequest(new Mesos.JenkinsSlave(TEST_JENKINS_SLAVE_NAME),0.2d,TEST_JENKINS_SLAVE_MEM,mesosSlaveInfo);
+    }
 
+    private JenkinsScheduler.Request mockMesosRequest(
+            Boolean useDocker,
+            Boolean useCustomDockerCommandShell,
+            String customDockerCommandShell) throws Descriptor.FormException {
+        Mesos.SlaveResult slaveResult = Mockito.mock(Mesos.SlaveResult.class);
+        return jenkinsScheduler.new Request(mockSlaveRequest(useDocker, useCustomDockerCommandShell,
+            customDockerCommandShell), slaveResult);
```
Comment (vinodkone) [inline src/test/java/org/jenkinsci/plugins/mesos/JenkinsSchedulerTest.java]: put 'mockSlaveRequest()' on next line and 'slaveResult' on the line after.

File: src/test/java/org/jenkinsci/plugins/mesos/JenkinsSchedulerTest.java [hunk 12]
```diff
@@ -146,6 +149,45 @@ public void run() {
...
+
+        SchedulerDriver driver = Mockito.mock(SchedulerDriver.class);
+        Mockito.when(mesosCloud.getRejectOfferDurationDouble()).thenReturn((double) 120000);
+        jenkinsScheduler.resourceOffers(driver, offers);
+        Mockito.verify(driver).declineOffer(offer.getId());
+        Mockito.verify(driver, Mockito.never()).declineOffer(offer.getId(), Protos.Filters.newBuilder().setRefuseSeconds(120000).build());
+    }
+
+    @Test
+    public void testReviveOffersWhenAddingSlaveRequest() throws Exception {
```
Comment (vinodkone) [inline src/test/java/org/jenkinsci/plugins/mesos/JenkinsSchedulerTest.java]: thanks for the tests! :+1:

**Comment by vinodkone** [comment]
Note that we exposed a new call in mesos called suppressOffers() that can suppress offers altogether instead of doing it one offer at a time as you did here. How about leveraging that instead?

**Comment by felixb** [comment]
I worked in most of your comments.

1. I'm still declining single offers. Because the new API call was introduced in mesos 0.25, We are still running 0.24.
2. I found, that the minimum `declineOfferDuration` in the jelly config is more or less a hint only. It's still possible for a user to set values below the minimum.

**Comment by felixb** [comment]
I just rebased to current master.
Any objections to merging this?
